include/boost/geometry/algorithms/detail/overlay/sort_by_side.hpp

```diff
@@ -291,20 +317,34 @@ public :
         geometry::copy_segment_points<Reverse1, Reverse2>(geometry1, geometry2,
                 op.seg_id, point1, point2, point3);
         Point const& point_to = op.fraction.is_one() ? point3 : point2;
-        add_segment(turn_index, op_index, point1, point_to, op.operation, op.seg_id, is_origin);
+
+        int offset = 0;
+
+        // If the point is in the neighbourhood (the limit itself is not important),
+        // then take a point (or more) further back.
+        // The limit of offset avoids theoretical infinite loops. In practice it currently
+        // walks max 1 point back in all cases.
+        while (approximately_equals(point1, turn.point, 1.0e-6) && offset > -10)
+        {
```

Comment:
This loop is the fix.

Reply:
What is this number `1.0e-6`? Will it work for `float` too? Will it work for big point coordinates when the closest next or previous floating point number is greater than `1.0e-6` ? Shouldn't it be scaled by coordinates like epsilon or maybe even based on machine epsilon?

Reply:
I wrote that in the comments. Machine epsilon won't work (I tried), it's too small and will still result in non-distinguishable sides (those are only indirectly based on these coordinates) - using them for side calculations afterwards.

Your concern about big coordinates might be valid, I have to check that.

Reply:
I mean **based on** machine epsilon. So taking both of my remarks into account e.g.:
```
n * epsilon * max_coordinate(turn.point)
```
where `n` is your constant for tweaking. As an alternative `approximately_equals` could check relatively to coordinates because AFAIU right now `1.0e-6` is absolute value. So it could take epsilon-based-error, e.g. `n * epsilon` and then internally use this value together with each one coordinate to check for approximate equality. So use separate error-thresholds for each coordinate, basically similar to `equals()` but with epsilon multiplied by some factor.

Side note: there are only 3 or 4 values of machine epsilon used in Boost.Geometry for a given machine. One for each floating point number or `0` for integers and user-defined types. So the above would not work well for user-defined floating point types unless `numeric_limits` are specialized for such type. And this is not always possible.

Reply:
Thanks @awulkiew !
I get the point and will apply it as you suggest.

Reply:
I'm afraid this doesn't work for the range of values I'm using.

For `double`: 1.0e9 * epsilon is 2.22045e-07, this is around the value I need. But it doesn't translate to other values.
For `float`: 1.0e9 * epsilon is 119.209 and of course we can't work with that limit...
For `long double`: 1.0e9 * epsilon is 1.0842e-10, this works still fine.

I can use it if I use the double epsilon for float, which is kind of awkward.

My implementation (but I won't push it):
```
    // Returns true if two points are approximately equal, tuned by a giga-epsilon constant
    // (if constant is 1.0, for type double, the boundary is about 1.0e-6)
    template <typename Point1, typename Point2, typename T>
    static inline bool approximately_equals(Point1 const& a, Point2 const& b, T giga_epsilon)
    {
        // Including distance would introduce cyclic dependencies.
        using coor_t = typename select_coordinate_type<Point1, Point2>::type;
        constexpr coor_t epsilon = std::numeric_limits<coor_t>::epsilon();

        coor_t const& a0 = geometry::get<0>(a);
        coor_t const& b0 = geometry::get<0>(b);
        coor_t const& a1 = geometry::get<1>(a);
        coor_t const& b1 = geometry::get<1>(b);
        coor_t const one = 1.0;
        auto const c = math::detail::greatest(a0, b0, a1, b1, one);
        auto const limit = 1.0e9 * giga_epsilon * epsilon * c;

        return std::abs(a0 - b0) <= limit && std::abs(a1 - b1) <= limit;
    }

```

And I call it with 
`while (approximately_equals(point1, turn.point, 1.0) && offset > -10)`

Reply:
OK, using a `min` I can fix it. I will submit this.

Reply:
:heavy_check_mark:

Reply:
OK, so now `limit` is in the order of `1.0e-3`, `1.0e-7`, `1.0e-10`, for `float`, `double`, `long double` resp. 

Does the current solution work well (or better than the original with fixed `1.0e-6`) for large coordinates? What about user defined floating point numbers? 

A similar question, is `limit_giga_epsilon` ever needed to be different than `1.0`?

Reply:
No, the solution doesn't work better, it works about the same (there's one case different for long double which I still have to investigate but I think it's not really related).

However, I didn't test with huge coordinates (maybe I should) or very very small coordinates. These testcases have values in the range of -100 .. 100. And Adam pointed out that it might turn out differently for other ranges, and therefore this solution is probably better.

About the limit: `1.0` works fine. But the limit is not crucial here: 2.0 will also work fine. Because it walks forwards or backwards to the next coordinate, and that will be fine unless there is a turn within the range of that limit...